### ISSUE TITLE
Index stories table by sources_id?

### ISSUE THREAD

**Issue by philbudne** [mediacloud/backup-rss-fetcher#26]
web search (mcweb) feed watcher makes copious use of `/api/sources/NNN/stories/fetched-by-day` endpoint, which generates queries like:
```
SELECT CAST(stories.fetched_at AS DATE) AS day, count(*) AS stories
FROM stories
WHERE stories.sources_id = $1::BIGINT GROUP BY CAST(stories.fetched_at AS DATE) ORDER BY day
```
the stories table currently has the following indices:
```
Indexes:
    "stories_pkey" PRIMARY KEY, btree (id)
    "stories_domain" btree (domain)
    "stories_feed_id" btree (feed_id)
    "stories_fetched_at" btree (fetched_at)
    "stories_published_at" btree (published_at)
    "unique_story_title" btree (normalized_title_hash, sources_id)
    "unique_story_url" UNIQUE, btree (normalized_url)
```
So each query has to scan the entire table, which has on the order of 45 million rows.

**Comment by philbudne** [comment]
From logs:
```
2023-10-17 02:52:28,017 | INFO | uvicorn.access | 172.19.0.7:53918 - "GET /api/sources/282703/stories/fetched-by-day HTTP/1.1" 200
2023-10-17 02:52:32,683 | INFO | server.util | endpoint: server.sources.fetch_source_stories_fetched_by_day, status: OK, duration: 4.6495983600616455 sec
2023-10-17 02:52:32,684 | INFO | uvicorn.access | 172.19.0.7:53918 - "GET /api/sources/282705/stories/fetched-by-day HTTP/1.1" 200
2023-10-17 02:52:37,626 | INFO | server.util | endpoint: server.sources.fetch_source_stories_fetched_by_day, status: OK, duration: 4.933398485183716 sec
2023-10-17 02:52:37,633 | INFO | uvicorn.access | 172.19.0.7:53918 - "GET /api/sources/282706/stories/fetched-by-day HTTP/1.1" 200
2023-10-17 02:52:42,604 | INFO | server.util | endpoint: server.sources.fetch_source_stories_fetched_by_day, status: OK, duration: 4.952731132507324 sec
2023-10-17 02:52:42,605 | INFO | uvicorn.access | 172.19.0.7:53918 - "GET /api/sources/20563/stories/fetched-by-day HTTP/1.1" 200
2023-10-17 02:52:47,474 | INFO | server.util | endpoint: server.sources.fetch_source_stories_fetched_by_day, status: OK, duration: 4.851433038711548 sec
2023-10-17 02:52:47,477 | INFO | uvicorn.access | 172.19.0.7:53918 - "GET /api/sources/20569/stories/fetched-by-day HTTP/1.1" 200
2023-10-17 02:52:52,322 | INFO | server.util | endpoint: server.sources.fetch_source_stories_fetched_by_day, status: OK, duration: 4.832204580307007 sec
2023-10-17 02:52:52,323 | INFO | uvicorn.access | 172.19.0.7:53918 - "GET /api/sources/282715/stories/fetched-by-day HTTP/1.1" 200
2023-10-17 02:52:57,192 | INFO | server.util | endpoint: server.sources.fetch_source_stories_fetched_by_day, status: OK, duration: 4.851716041564941 sec
2023-10-17 02:52:57,194 | INFO | uvicorn.access | 172.19.0.7:53918 - "GET /api/sources/282723/stories/fetched-by-day HTTP/1.1" 200
2023-10-17 02:53:02,104 | INFO | server.util | endpoint: server.sources.fetch_source_stories_fetched_by_day, status: OK, duration: 4.892181158065796 sec
2023-10-17 02:53:02,105 | INFO | uvicorn.access | 172.19.0.7:53918 - "GET /api/sources/20579/stories/fetched-by-day HTTP/1.1" 200
2023-10-17 02:53:06,869 | INFO | server.util | endpoint: server.sources.fetch_source_stories_fetched_by_day, status: OK, duration: 4.757479190826416 sec
2023-10-17 02:53:06,870 | INFO | uvicorn.access | 172.19.0.7:53918 - "GET /api/sources/20580/stories/fetched-by-day HTTP/1.1" 200
2023-10-17 02:53:11,481 | INFO | server.util | endpoint: server.sources.fetch_source_stories_fetched_by_day, status: OK, duration: 4.602452278137207 sec
2023-10-17 02:53:11,482 | INFO | uvicorn.access | 172.19.0.7:53918 - "GET /api/sources/20581/stories/fetched-by-day HTTP/1.1" 200
2023-10-17 02:53:16,318 | INFO | server.util | endpoint: server.sources.fetch_source_stories_fetched_by_day, status: OK, duration: 4.82752537727356 sec
```

**Comment by philbudne** [comment]
The feed watcher can issue 18K queries, at 5 seconds each, than can multiply out to 25 hours!!!

**Comment by philbudne** [comment]
Implemented in v0.16.4 https://github.com/mediacloud/rss-fetcher/commit/9a3c8308e6cec7bc6ca3348d5c4c53f7c7ccd4c7 after testing that migration ran quickly.  Database size is still nominally 58G.

**Comment by philbudne** [comment]
Looks like (on average) at 700x speedup (from average of 5.3 sec to average of 0.006931 sec)

### BEFORE:
```
2024-08-09 04:36:04,017 | INFO | server.util | endpoint: server.sources.fetch_source_stories_fetched_by_day, status: OK, duration: 5.307758808135986 sec
2024-08-09 04:36:04,018 | INFO | uvicorn.access | 172.19.0.3:48126 - "GET /api/sources/131070/stories/fetched-by-day HTTP/1.1" 200
--
2024-08-09 04:36:08,821 | INFO | server.util | endpoint: server.sources.fetch_source_stories_fetched_by_day, status: OK, duration: 4.795267343521118 sec
2024-08-09 04:36:08,822 | INFO | uvicorn.access | 172.19.0.3:48126 - "GET /api/sources/655359/stories/fetched-by-day HTTP/1.1" 200
--
2024-08-15 13:12:11,931 | INFO | server.util | endpoint: server.sources.fetch_source_stories_fetched_by_day, status: OK, duration: 6.822664976119995 sec
2024-08-15 13:12:11,935 | INFO | uvicorn.access | 172.19.0.32:51134 - "GET /api/sources/262319/stories/fetched-by-day HTTP/1.1" 200
--
2024-08-15 14:34:50,388 | INFO | server.util | endpoint: server.sources.fetch_source_stories_fetched_by_day, status: OK, duration: 6.648968696594238 sec
2024-08-15 14:34:50,393 | INFO | uvicorn.access | 172.19.0.32:51134 - "GET /api/sources/656646/stories/fetched-by-day HTTP/1.1" 200
--
2024-08-15 21:23:29,492 | INFO | server.util | endpoint: server.sources.fetch_source_stories_fetched_by_day, status: OK, duration: 4.440416574478149 sec
2024-08-15 21:23:29,493 | INFO | uvicorn.access | 172.19.0.32:51134 - "GET /api/sources/662687/stories/fetched-by-day HTTP/1.1" 200


Total for 3083 calls 8/9/2024:
pbudne@tarbell:~/rss-fetcher/prod/logs$ grep fetch_source_stories_fetched_by_day server.web.1.log.2024-08-09 | awk '{ total += $13 } END { print total }'
16533.2
average: 5.3 sec

Total for 2381 calls 8/16/20204:
pbudne@tarbell:~/rss-fetcher/prod/logs$ grep -v '2024-08-16 19:'  server.web.1.log | grep fetch_source_stories_fetched_by_day | awk '{ total += $13 } END { print total }'
12728.2
average: 5.3 sec
```

### AFTER:
```
2024-08-16 19:23:31,130 | INFO | server.util | endpoint: server.sources.fetch_source_stories_fetched_by_day, status: OK, duration: 0.5617661476135254 sec
2024-08-16 19:23:31,135 | INFO | uvicorn.access | 172.19.0.32:45522 - "GET /api/sources/262319/stories/fetched-by-day HTTP/1.1" 200
--
2024-08-16 19:23:45,802 | INFO | server.util | endpoint: server.sources.fetch_source_stories_fetched_by_day, status: OK, duration: 0.4394223690032959 sec
2024-08-16 19:23:45,808 | INFO | uvicorn.access | 172.19.0.32:45522 - "GET /api/sources/656646/stories/fetched-by-day HTTP/1.1" 200
--
2024-08-16 19:24:46,831 | INFO | uvicorn.access | 172.19.0.32:45522 - "GET /api/sources/662687/stories/fetched-by-day HTTP/1.1" 200
2024-08-16 19:24:46,842 | INFO | server.util | endpoint: server.sources.fetch_source_stories_fetched_by_day, status: OK, duration: 0.0037703514099121094 sec
--
2024-08-16 19:33:40,087 | INFO | server.util | endpoint: server.sources.fetch_source_stories_fetched_by_day, status: OK, duration: 0.002958536148071289 sec
2024-08-16 19:33:40,087 | INFO | uvicorn.access | 172.19.0.32:45522 - "GET /api/sources/131070/stories/fetched-by-day HTTP/1.1" 200
--
2024-08-16 19:33:40,094 | INFO | server.util | endpoint: server.sources.fetch_source_stories_fetched_by_day, status: OK, duration: 0.003149271011352539 sec
2024-08-16 19:33:40,094 | INFO | uvicorn.access | 172.19.0.32:45522 - "GET /api/sources/655359/stories/fetched-by-day HTTP/1.1" 200

Total for 39108 calls 8/16/2024:
pbudne@tarbell:~/rss-fetcher/prod/logs$ grep '2024-08-16 19:'  server.web.1.log | grep fetch_source_stories_fetched_by_day | wc -l
39108
pbudne@tarbell:~/rss-fetcher/prod/logs$ grep '2024-08-16 19:'  server.web.1.log | grep fetch_source_stories_fetched_by_day | awk '{ total += $13 } END { print total }'
271.06
```
